# Handing missing values:


I am going to follow the steps mentioned below in this notebook:

* [Take a first look at the data](#Take-a-first-look-at-the-data)
* [See how many missing data points we have](#See-how-many-missing-data-points-we-have)
* [Figure out why the data is missing](#Figure-out-why-the-data-is-missing)
* [Drop missing values](#Drop-missing-values)
* [Filling in missing values](#Filling-in-missing-values)

Let's get started!

## Take a first look at the data
________

The first thing we'll need to do is load in the libraries and datasets we'll be using. For today, I'll be using a dataset of books with different specifications.


In [1]:
# modules we'll use
import pandas as pd
import numpy as np

# read in all our data
books_data = pd.read_csv("BL-Flickr-Images-Book.csv")


The first thing I do when I get a new dataset is take a look at some of it. This lets me see that it all read in correctly and get an idea of what's going on with the data. In this case, I'm looking to see if I see any missing values, which will be reprsented with `NaN` or `None`.

In [2]:
# look at a few rows of the books_data file. I can see a handful of missing data already!
books_data.sample(5)

Identifier Edition Statement Place of Publication Date of Publication  \
2942     1445091               NaN            Cape Town                1855   
365       151917               NaN                Paris                1858   
6590     3230083               NaN     Château-Thierry                1894   
8190     3993027               NaN    Frankfurt am Main                1838   
880       434723               NaN               London                1787   

                   Publisher  \
2942        S. Solomon & Co.   
365                      NaN   
6590                     NaN   
8190                     NaN   
880   G. G. J. & J. Robinson   

                                                  Title  \
2942  Sunshine and Cloud; or, light thrown on a dark...   
365   Les Voyages de Améric Vespuce au compte de l'...   
6590  Un Coin de la Champagne et du Valois au XVIIe ...   
8190  Zur Sprach- und Geschichtsforschung der neuest...   
880   The Midnight Hour. A comedy, in three acts. Fr...   

                                                 Author  \
2942                                 GODLONTON, Robert.   
365              AVEZAC-MACAYA, Marie Armand Pascal d'.   
6590                                        SALESSE, I.   
8190       XYLANDER, Karl August Anton Aloys Josef von.   
880   BOURLIN, Antoine Jean - calling himself Dumaniant   

                                           Contributors  Corporate Author  \
2942                  STOCKENSTROM, Andries - Sir, Bart               NaN   
365   VARNHAGEN, Francisco Adolfo de - Viscount de P...               NaN   
6590                                  HÉRICART, Marie.               NaN   
8190                      SCHOTT, Wilhelm - Orientalist               NaN   
880                                      Inchbald - Mrs               NaN   

      Corporate Contributors Former owner  Engraver Issuance type  \
2942                     NaN          NaN       NaN   monographic   
365                      NaN          NaN       NaN   monographic   
6590                     NaN          NaN       NaN   monographic   
8190                     NaN          NaN       NaN   monographic   
880                      NaN          NaN       NaN   monographic   

                                             Flickr URL  \
2942  http://www.flickr.com/photos/britishlibrary/ta...   
365   http://www.flickr.com/photos/britishlibrary/ta...   
6590  http://www.flickr.com/photos/britishlibrary/ta...   
8190  http://www.flickr.com/photos/britishlibrary/ta...   
880   http://www.flickr.com/photos/britishlibrary/ta...   

                                             Shelfmarks  
2942               British Library HMNTS 9004.h.30.(3.)  
365                    British Library HMNTS 10480.e.9.  
6590                  British Library HMNTS 010171.i.2.  
8190  British Library HMNTS 9007.dd.3.|British Libra...  
880   British Library HMNTS 643.i.11.(2.)|British Li...

## See how many missing data points we have
___

Ok, now we know that we do have some missing values. Let's see how many we have in each column. 

In [3]:
# get the number of missing data points per column
missing_values_count = books_data.isnull().sum()
missing_values_count

Identifier                   0
Edition Statement         7514
Place of Publication         0
Date of Publication        181
Publisher                 4195
Title                        0
Author                    1778
Contributors                 0
Corporate Author          8287
Corporate Contributors    8287
Former owner              8286
Engraver                  8287
Issuance type                0
Flickr URL                   0
Shelfmarks                   0
dtype: int64

That seems like a lot! It might be helpful to see what percentage of the values in our dataset were missing to give us a better sense of the scale of this problem:

In [4]:
# how many total missing values do we have?
total_cells = np.product(books_data.shape)
total_missing = missing_values_count.sum()

# percent of data that is missing
(total_missing/total_cells) * 100

37.66139736937372

Wow, more than a quarter of the cells in this dataset are empty! In the next step, we're going to take a closer look at some of the columns with missing values and try to figure out what might be going on with them.

## Figure out why the data is missing
____
 

Before dealing with missing values, we'll need to use our intution to figure out why the value is missing. One important question that we can ask ourselves is **if the values of a column are missing becuase they weren't recorded or because they don't exist**?

If a value is missing because it doens't exist (like the height of the oldest child of someone who doesn't have any children) then it doesn't make sense to try and guess what it might be. These values we probalby do want to keep as NaN. On the other hand, if a value is missing becuase it wasn't recorded, then we can try to guess what it might have been based on the other values in that column and row which we call as `imputation`.



In [5]:
# look at the # of missing points in all columns
missing_values_count

Identifier                   0
Edition Statement         7514
Place of Publication         0
Date of Publication        181
Publisher                 4195
Title                        0
Author                    1778
Contributors                 0
Corporate Author          8287
Corporate Contributors    8287
Former owner              8286
Engraver                  8287
Issuance type                0
Flickr URL                   0
Shelfmarks                   0
dtype: int64

In [6]:
books_data.shape

(8287, 15)

## Drop missing values
____

**Let us handle missing values for different columns one by one.** 
>We can see that for `Edition Statement` approx. 90% of the data is missing since not every book has a new edition so it is fair to remove this column.
Similarly for columns like `Corporate Author`,`Corporate Contributors`,`Former owner`and `Engraver`, 100% of the data is missing so we will remove these columns as well. 



In [7]:
books_data.drop(['Edition Statement','Corporate Author','Corporate Contributors','Former owner','Engraver'], axis = 1, inplace = True) 

## Filling in missing values 
_______
 Filling in the missing values is called **Imputation.**

>`Author` column has 1778 missing values. Since Authors column would have a different specific value each time, we will treat these missing values as a specific categorty. Hence we will replace the NaN values with a Na value. 
Similarly, the same relacement could be done for the Publisher column as well. 

In [47]:
# Replacing NaN values with Na for Author 
books_data['Author'].replace(to_replace= np.nan, value="Na", inplace=True)

In [10]:
books_data.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                  Contributors Issuance type  \
0      A. A.                               FORBES, Walter.   monographic   
1  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
2  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
3  A., E. S.                   Appleyard, Ernest Silvanus.   monographic   
4  A., E. S.                           BROOME, John Henry.   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [48]:
# Replacing NaN values with Na for Publisher
books_data['Publisher'].fillna("Na", inplace = True)

>For `Date of Publication` column we can impute some values for NaN values. But before doing that, we need to clean the column as it contains certain unwanted characters. Hence we will first remove these unwanted characters and then do the `imputation`.

In [49]:
#Cleaning Date of Publication
unwanted_characters = ['[', ',', '-','.',']','?','/']

def clean_dates(dop):
    dop = str(dop)
    if dop.startswith('[') or dop == 'nan':
        return 'NaN'
    for character in unwanted_characters:
        if character in dop:
            character_index = dop.find(character)
            dop = dop[:character_index]
    return dop

books_data['Date of Publication'] = books_data['Date of Publication'].apply(clean_dates)
books_data.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London               1879    
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                  Contributors Issuance type  \
0      A. A.                               FORBES, Walter.   monographic   
1  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
2  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
3  A., E. S.                   Appleyard, Ernest Silvanus.   monographic   
4  A., E. S.                           BROOME, John Henry.   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

>Now that we have removed the unwanted characters from the `Date of Publication` column, we can impute the NaN values. One way to do that is to sort the values in ascending order and then fill the NaN values with the values occuring just after them.(This makes a lot of sense for datasets where the observations have some sort of logical order to them.)

In [50]:
# Arranging the values in ascending order
sorted_df = books_data.sort_values(by = 'Date of Publication', ascending = True)
# replace all NaN's with the values that comes directly after it in the same column, 
sorted_df.fillna(method = 'bfill', axis=0)

Identifier                               Place of Publication  \
1081      516336                                             London   
5944     2938347                                              Paris   
2934     1442148                                           Lond[on]   
2817     1376528                                             London   
7778     3823617                                             London   
4852     2396288                                             London   
895       439737                                             London   
6528     3212310                                             London   
2035      896061                                             London   
1453      660893                                             London   
7326     3582834                                             London   
3515     1704542                                             London   
4613     2299790                                             London   
6160     3022543                                             London   
3094     1509795                                             London   
3757     1823663                                             London   
1080      515955                                             London   
7776     3823253                                             London   
6161     3022545                                             London   
6898     3370994                                             London   
2604     1253024                                             London   
1454      660943                                             London   
2479     1186704                                             London   
7053     3442119                                             London   
2603     1253003                                             London   
564       241254                                             London   
5037     2486393                                             London   
3187     1553618                                             Oxford   
4984     2462653                                             London   
4985     2462657                                           London,]   
...          ...                                                ...   
6489     3191783                                           New York   
6487     3191349                                           New York   
501       207672                                             London   
3673     1769127                                             Zurich   
3607     1751091                                             London   
6548     3214222                                    London, [1874.]   
6550     3214363              pp. 23. T. Fisher Unwin: London, 1883   
6637     3263773        pp. 80. Andrew Churchman: Hammersmith, 1890   
3621     1759168                                            Wrexham   
3625     1760060                                            Chester   
6623     3257016                                       London, 1838   
6622     3256861                                     New York, 1866   
459       179539                                             London   
6617     3254078                                             London   
6615     3252024                   pp. 46. E. Mathews: London, 1898   
464       182354                                             London   
6608     3245710                       pp. 16. C. Fox: London, 1835   
6603     3242628                                         Birmingham   
467       183576                                             London   
6583     3226247                                             London   
6581     3226025                                             London   
3644     1761576  A[ppledore]. P[rivate]. P[ress: New-Haven, Con...   
3650     1762496                                     Great Yarmouth   
1247      559884                                             London   
2374     1045858              

>Now that we have dealt with all the missing values of different columns, let us check it once.

In [44]:
sorted_df.isnull().sum()

Identifier              0
Place of Publication    0
Date of Publication     0
Publisher               0
Title                   0
Author                  0
Contributors            0
Issuance type           0
Flickr URL              0
Shelfmarks              0
dtype: int64